openlibrary/tests/fastapi/test_search.py

```diff
@@ -0,0 +1,218 @@
+            ],
+        }
+        yield mock
+
+
+def search(client, **params):
+    """Helper function to make search requests with query parameters.
+
+    This helper provides a cleaner interface for making search requests
+    by accepting keyword arguments directly instead of building URL strings.
+    """
+    from urllib.parse import urlencode
+
+    query_string = urlencode(params, doseq=True)
+    return client.get(f'/search.json?{query_string}')
+
+
+class TestSearchEndpoint:
+    """Tests for the /search.json endpoint."""
+
+    def test_search_uses_query_param(self, client, mock_work_search):
+        """Test that the search endpoint uses the 'q' query parameter."""
+        mock_work_search.return_value = {
+            'numFound': 1,
+            'start': 0,
+            'docs': [{'key': '/works/OL1W', 'title': 'The Lord of the Rings'}],
+        }
+
+        response = search(client, q='lord of the rings')
+
+        assert response.status_code == 200
+        data = response.json()
+
+        # Verify the response contains the query
+        assert data['q'] == 'lord of the rings'
+        assert 'docs' in data
+        assert 'numFound' in data
+
+        # Verify work_search_async was called with the query
+        mock_work_search.assert_called_once()
+        call_args = mock_work_search.call_args
+        query_arg = call_args[0][0]  # First positional argument
+        assert query_arg['q'] == 'lord of the rings'
+
+    def test_search_uses_query_alias_param(self, client, mock_work_search):
+        """Test that the search endpoint uses the 'query' query parameter (alias for full JSON query)."""
+        mock_work_search.return_value = {
+            'numFound': 1,
+            'start': 0,
+            'docs': [{'key': '/works/OL1W', 'title': 'Test'}],
+        }
+
+        # The 'query' param should accept a full JSON-encoded Solr query
+        query_dict = {'q': 'test', 'author': 'Tolkien'}
+        response = search(client, query=json.dumps(query_dict))
+
+        assert response.status_code == 200
+
+        # Verify work_search_async was called with the parsed query
+        mock_work_search.assert_called_once()
+        call_args = mock_work_search.call_args
+        query_arg = call_args[0][0]
+        assert query_arg == query_dict
+
+    @pytest.mark.parametrize(
+        ('params', 'expected_kwargs'),
+        [
+            (
+                {'q': 'test', 'page': 3, 'limit': 10},
+                {'page': 3, 'limit': 10, 'offset': None},
+            ),
+            (
+                {'q': 'test', 'offset': 50, 'limit': 25},
+                {'offset': 50, 'limit': 25, 'page': None},
+            ),
+            (
+                {'q': 'test', 'page': 5, 'offset': 30, 'limit': 10},
+                {'offset': 30, 'limit': 10, 'page': None},
+            ),
+            (
+                {'q': 'test'},
+                {'limit': 100, 'page': 1, 'offset': None},
+            ),
```

Comment:
Adding a test for limit=0 would be useful ; this feature has been accidentally removed by folks in that past at various points :P